### ISSUE TITLE
Issues loading non ESM bundles in browsers with polyfills. 

### ISSUE THREAD

**Issue by seivan** [rollup/rollup#3611]
### Expected Behavior / Situation
To match these requirements: 
1) Deploy a working frontend application
2)  Make it work on browsers that don't support native ES Modules.
3) Use code splitting (this excludes `UMD` and `IIFE`) 
4) Allow chunking large dependencies and/or smaller less updated dependencies
5) Be as unobtrusive as possible in the code defined inside `./src`, it's ok with conditionals in the `index.html` but not actual code base . 
6) Do not rely on external CDNs.
7) Not use Babel - We actually use `Babel` & `CoreJS+entry` to generate the required polyfills, but that's pre-build step and that's it. 

Now as I understand it, to get all those requirements listed above you need to use `SystemJS` or `CJS`  and that would require some polyfill like 
https://github.com/systemjs/systemjs
or
https://github.com/luciVuc/browser-cjs

I am open to suggestion for other module loaders, as long as they are small and not intrusive. 

However those libraries depend on other polyfills like `Promise` or `Fetch`
So my take was something like a `shims.ts` file as an `entry/input` that uses `core-js@3`

Now assume we decide to go with `SystemJS`: 

```ts
//Inside shims.ts
import "core-js/modules/es.symbol";
import "core-js/modules/es.symbol.description";
import "core-js/modules/es.symbol.async-iterator";
import "core-js/modules/es.symbol.has-instance";
import "core-js/modules/es.symbol.is-concat-spreadable";
import "core-js/modules/es.symbol.iterator";
import "core-js/modules/es.symbol.match";
//..... even more
```
I would like to be able to load the output of that as a script tag. 

### Actual Behavior / Situation
However....
The output of `shim.ts` file gets wrapped in a `System.register()` and keep in mind, that `SystemJS` has yet to be loaded! 

There are some other issues, like for instance I rather *NOT* do this .
```ts
        copy({
            targets: [
                { src: ['node_modules/systemjs/dist/s.js', 'node_modules/unfetch/polyfill/index.js'], dest: 'dist/' },

            ]
        }),
```
I would like to be able to import these directly in my typescript code, because I want to support content hashing for cache busting purposes and make it actually belongs in code instead of a build step - also it seems kinda like a hack. There is no reason why I can't import these files into my actual code. But I can't since Rollup doesn't let me control the output of a few selected entry files. 

### Modification Proposal

No idea, actually. 
I am open to suggestions, and maybe I am doing something wrong, well most likely, but it doesn't change the requirement I listed above as I find them to be decent practices. 

What I really would like is something like this 

```ts
//Inside shims.ts
import "core-js/modules/es.symbol";
import "core-js/modules/es.symbol.description";
import "core-js/modules/es.symbol.async-iterator";
import "core-js/modules/es.symbol.has-instance";
import "core-js/modules/es.symbol.is-concat-spreadable";
import "core-js/modules/es.symbol.iterator";
import "core-js/modules/es.symbol.match";
import "unfetch/polyfill"
import s from "systemjs/dist/s"
//An adding the bundle/module loaders
```
This shoud load all the shims needed make sure they are available globally, then load the shims needed for `SystemJS` and which is also set globally.

Then the rest of code can just work as normal.
```html
<script src="shims.js"></script>
<script>
System.import("vendor.js")
System.import("react.js")
System.import("app.js")
</script>
```

or I would prefer if possible: 

```html
<script src="shims.js"></script>
<script src="vendor.js"></script>
<script src="react.js"></script>
<script src="app.js"></script>
```

TL;DR Is there a way to avoid wrapping an `entry/input` in `System.register()` and just have the code as is.
If it sets global/window variables, so be it. That's the point of a global shim like `SystemJS` and `core-js`. They aren't imported in individual files/components as use them like say `React` or `RxJS` is.

**Comment by guybedford** [comment]
SystemJS is designed to be loaded as a script, and any extras, polyfills or additional functionality can certainly be inlined along with SystemJS - if you really wanted to do this with a RollupJS build it would have to be an IIFE module for the boot or alternatively a simple concatenation process.

There is also no need to load polyfills in environments that don't need it so having, so SystemJS does include an advised section here to use a feature detection with `document.write` in order to optimize for modern browser performance - https://github.com/systemjs/systemjs#polyfills-for-older-browsers.

Core-js polyfills are much more granular than the above and may make more sense to include in the actual modular build though.

If you have an entry point that needs certain polyfills for the main modular build (after getting the boot loader worked out), then import order can be relied on in RollupJS as a way to ensure polyfills load first:

```js
import "core-js/modules/es.symbol";
import app from './app.js';
app.load();
```

But I would not suggest unifying this treatment for the fetch and Promise polyfills in SystemJS since that boot polyfill phase is far more critical and worth giving special attention to.

Hope that answers some of your questions.

**Comment by seivan** [comment]
> SystemJS is designed to be loaded as a script, and any extras, polyfills or additional functionality can certainly be inlined along with SystemJS - if you really wanted to do this with a RollupJS build it would have to be an IIFE module for the boot or alternatively a simple concatenation process.

This is interesting, I was under the impression that IIFE modules were isolated, so you're saying- and correct me if I am wrong: 

I can create something like 

```ts
//fat_polyfill_es5.ts
import "everything-from-core-js"
import "unfetch/polyfill" 
import s from "systemjs/dist/s" //Any suggestions on this? 
```

> There is also no need to load polyfills in environments that don't need it so having, so SystemJS does include an advised section here to use a feature detection with `document.write` in order to optimize for modern browser performance - https://github.com/systemjs/systemjs#polyfills-for-older-browsers.
> 
> Core-js polyfills are much more granular than the above and may make more sense to include in the actual modular build though.

Just to be clear, any environment where `SystemJS` is required, will get a fat polyfill from `core-js` - since we deem them to be legacy browsers and aren't optimizing for more than allowing lazy loaded components (code splitting), anything else is a bonus. Right now we're using `IIFE` for legacy browsers, which Rollup doesn't allow code splitting. 


> If you have an entry point that needs certain polyfills for the main modular build (after getting the boot loader worked out), then import order can be relied on in RollupJS as a way to ensure polyfills load first:
> 
> ```js
> import "core-js/modules/es.symbol";
> import app from './app.js';
> app.load();
> ```
> 
> But I would not suggest unifying this treatment for the fetch and Promise polyfills in SystemJS since that boot polyfill phase is far more critical and worth giving special attention to.

Not sure I follow, `SystemJS` itself required `fetch` polyfill and I might as well include the rest of `ES5` polyfills in there before loading `SystemJS`. 
Once all that is setup, add it as a single script tag before anything else is loaded after conditionally checking if it's needed within [inlined `<script>` tag](https://github.com/systemjs/systemjs/issues/2194#issuecomment-637937431)

> 
> Hope that answers some of your questions.

Thanks, I think it gave me an idea of doing separate builds.
First `IIFE` for `core-js+fetch+SystemJS` 
Second entire app as `SystemJS` with code splitting. 

If I understood your suggestion correctly?

**Comment by seivan** [comment]
@guybedford Oh wow your suggestion worked like a charm! 
The only think I had missed was that it was that loading via `script` tag isn't valid.
You actually have to do 
```html
    <script src="shims.js"></script>
    <script>
        
      System.import('./entry_application.js').then(function(m) {
      	console.log(m)
      })
    </script>
```

In fact, doing both that and setting the `script` tag will download it twice. 
I just tested this on IE11 and worked like a charm.

**Comment by seivan** [comment]
Closing this unless @guybedford has any input. 
The only issue is that this requires several build steps with rollup which makes it harder to setup a watch/hot code reload and get size/dependency feedback at a build step. 
But that's a separate issue with Rollup not allowing granular setups for entry and outputs in a single build step. 

Thanks again!

**Comment by LarsDenBakker** [comment]
For web apps I created https://github.com/open-wc/open-wc/tree/master/packages/rollup-plugin-html with https://github.com/open-wc/open-wc/tree/master/packages/rollup-plugin-polyfills-loader, which can create the correct loader and polyfills based on your browser requirements. 

The polyfills are not bundled, but loaded separately based on feature detection. Filenames are hashed so that you can do caching. Maybe it is useful to you.

**Comment by seivan** [comment]
@LarsDenBakker Wow `@open-wc/rollup-plugin-html` is perfect, thanks!
